crates/span/src/hygiene.rs

```diff
@@ -27,7 +27,10 @@ use crate::Edition;
 #[cfg(feature = "salsa")]
 #[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
 pub struct SyntaxContext(
-    salsa::Id,
+    /// # Invariant
+    ///
+    /// This is either a valid `salsa::Id` or a root `SyntaxContext`.
+    u32,
```

Comment:
Why do we need this change? Ideally in the future we will be able to use the salsa macros to define the type here and then this change is moot.

Reply:
I did it in order to encapsulate the unsafety of creating an `Id` well; previously it was already an `Id` and you could just hope nobody will ever use it incorrectly if it's a root `SyntaxContext` (that isn't really an `Id`), now the conversion to `Id` checks if it is a root.

If/when we convert this to Salsa macros this can go.